edalize/vivado.py

```diff
@@ -88,21 +120,43 @@ def configure_main(self):
             'has_xci'      : has_xci,
         }
 
-        self.render_template('vivado-project.tcl.j2',
-                             self.name+'.tcl',
-                             template_vars)
+        if self.synth_tool == 'yosys':
+            xdc_file = next((f for f in src_files if f.file_type == 'xdc'), None)
+            if xdc_file is not None:
+                xdc_file = xdc_file.name
+            self.render_template('vivado-project-yosys.tcl.j2',
+                                 self.name+'.tcl',
+                                 {'name' : self.name,
+                                  'tool_options' : self.tool_options})
+
+            self.render_template('vivado-run-yosys.tcl.j2',
+                                 self.name+"_run.tcl",
+                                 {'edif_file' : self.name + '.edif',
+                                  'xdc_file' : xdc_file,
+                                  'name' : self.name,
+                                  'tool_options' : self.tool_options})
+        else:
+            self.render_template('vivado-project.tcl.j2',
+                                 self.name+'.tcl',
+                                 template_vars)
+
+            self.render_template('vivado-run.tcl.j2',
+                                 self.name+"_run.tcl")
+
+            self.render_template('vivado-synth.tcl.j2',
+                                 self.name+"_synth.tcl")
+
+        vivado_settings = self.tool_options.get('vivado-settings', None)
+        vivado_command = "source {} && vivado".format(vivado_settings) if vivado_settings else "vivado"
```

Comment:
I've always assumed any required environment setup would be done outside of the tool. What's the use case for doing this in edalize?

Reply:
I think that sourcing, at least, the Vivado settings prior to the command invocation is better as it may happen that the environment created by Vivado alters some of the core libraries such as libc.

By doing like this, we can be sure that the environment of the caller is preserved and it is altered only in the time frame in which the command is executed.

Reply:
I like this approach. As @GCHQDeveloper560 says, this has traditionally been done outside of fusesoc/edalize. The main reason for that was to leave full control to the user and avoid any attempts at autodetecting these kind of things. But there are definitely cases where sourcing the scripts file would be better and I haven't thought of doing it with tool_options. This is pretty nice since we can do it on the command-line.